sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpMessageConverter.cs

```diff
@@ -315,26 +333,200 @@ public virtual PartitionProperties CreatePartitionPropertiesFromResponse(AmqpMes
         private static AmqpMessage BuildAmqpMessageFromEvent(EventData source,
                                                              string partitionKey)
         {
-            if (!MemoryMarshal.TryGetArray(source.EventBody.ToMemory(), out var bodySegment))
+            var sourceMessage = source.GetRawAmqpMessage();
+
+            var message = sourceMessage switch
             {
-                bodySegment = new ArraySegment<byte>(source.EventBody.ToArray());
+                _ when sourceMessage.Body.TryGetData(out var dataBody) => AmqpMessage.Create(TranslateDataBody(dataBody)),
+                _ when sourceMessage.Body.TryGetSequence(out var sequenceBody) => AmqpMessage.Create(TranslateSequenceBody(sequenceBody)),
+                _ when sourceMessage.Body.TryGetValue(out var valueBody) => AmqpMessage.Create(TranslateValueBody(valueBody)),
+                _ => AmqpMessage.Create(new Data { Value = new ArraySegment<byte>(Array.Empty<byte>()) })
+            };
+
+            // Header
+
+            if (sourceMessage.HasSection(AmqpMessageSection.Header))
+            {
+                if (sourceMessage.Header.DeliveryCount.HasValue)
```

Comment:
The AMQP library will trigger allocations for some properties when a `null` or `default` value is assigned.  This applies both to the section itself and to some individual properties.  For example, properties of type `AmqpSymbol?` will accept a `null` string value, but perform an implicit conversion to `new AmqpSymbol(null)` when the setter is invoked.  In addition to the allocation, this will also cause the message size to inflate unnecessarily, using more bandwidth.   

In order to ensure that we don't trigger unwanted allocations and keep consistency in our approach, the conversion is ensuring a value exists before invoking any setters for the AMQP transport types.